### ISSUE TITLE
bug: Timestamp values

### ISSUE THREAD

**Issue by Paganiniana** [capawesome-team/capacitor-firebase#474]
### Plugin(s)

- [ ] Analytics
- [ ] App
- [ ] App Check
- [ ] Authentication
- [ ] Crashlytics
- [X] Cloud Firestore
- [ ] Cloud Messaging
- [ ] Performance
- [ ] Remote Config

### Did you test the latest version?

- [X] I use the latest version

### Platform(s)

- [ ] Android
- [X] iOS
- [ ] Web

### Current behavior

If I read a document using the `FirebaseFirestore.getDocument` method, its `snapshot.data` object is completely missing any value that resolves as a Firestore `Timestamp` value. For example, a document of the following format in Firebase ...

```js
const docData = {
     author: "Some Name",
     description: "Some description",
     creationDate: <some timestamp>,
}
```

... is returned as ...

```js
const docData = {
     author: "Some Name",
     description: "Some description",
}
```

### Expected behavior

I would expect that any `snapshot.data` value has certain values in Firestore, it this method would return those values as part of its resolved object structure.

### Reproduction

https://github.com/Paganiniana/capawesome-bug-reproduction

### Steps to reproduce

(See the readme)
1. build project
2. run on local web server
3. run on an iOS device
4. observe difference in console logs

### Capacitor doctor

💊   Capacitor Doctor  💊 

Latest Dependencies:

  @capacitor/cli: 5.5.0
  @capacitor/core: 5.5.0
  @capacitor/android: 5.5.0
  @capacitor/ios: 5.5.0

Installed Dependencies:

  @capacitor/android: not installed
  @capacitor/cli: 5.5.0
  @capacitor/ios: 5.5.0
  @capacitor/core: 5.5.0

[success] iOS looking great! 👌

### Before submitting

- [X] I understand that incomplete issues (e.g. without reproduction) are closed.

**Comment by Paganiniana** [comment]
In poking around the codebase, I've narrowed down the problem to the `FirebaseFirestoreHelper.swift` file. It seems the `createJSValue` method returns `nil` for instances of `Timestamp`. I'm manually marking its type and passing up its value as a second/nano second pair. **_In the JS layer, I'm passing these to the timestamp constructor, making it work just like the web client._** If we could make these changes in `firestore/src/index.ts` and wrap the necessary read/write methods to pass this information back and forth from the native/web layers, that would be great!

![Screenshot 2023-10-23 at 4 03 50 PM](https://github.com/capawesome-team/capacitor-firebase/assets/11188759/7955a520-a55a-4a8b-8062-f966009c9774)

**Comment by robingenz** [comment]
Thank you for reporting this issue! Would you be willing to create a PR with your idea? I would then take a closer look.

**Comment by Paganiniana** [comment]
Sure thing, @robingenz . I'll see if I can get that together, this morning.

**Comment by SureshKumarToverto** [comment]
@robingenz Could you please provide any updates on this matter or offer an estimated timeframe for resolving the issue? Thank you. 

1. #564 
2. #475

**Comment by SureshKumarToverto** [comment]
@robingenz Could you please provide any updates on this matter or offer an estimated timeframe for resolving the issue? Thank you. 
1. #564 
2. #475

**Comment by robingenz** [comment]
@SureshKumarToverto As already mentioned in https://github.com/capawesome-team/capacitor-firebase/issues/564#issuecomment-1952422981, this is currently a bug in Capacitor (see https://github.com/ionic-team/capacitor/issues/7252). I have to wait until it's fixed.

**Comment by johnmckay-reward** [comment]
@robingenz does this same bug also affect geopoints? we store both timestamps and geopoints in our firestore db, everything comes back fine when testing in the browser, but when we compile into a native iOS app and test on the device, neither timestamps/geopoints come back

**Comment by robingenz** [comment]
@johnmckay-reward No, this should be another issue. Please create a separate bug report.

**Comment by mdean808** [comment]
Is there a workaround for this bug while waiting for the capacitor team to fix https://github.com/ionic-team/capacitor/issues/7252, or should I switch to storing my dates as strings as opposed to the Timestamp type in firestore to avoid this issue?

**Comment by robingenz** [comment]
No, there is currently no workaround. I have another idea on how to solve this, but of course I would prefer the problem to be solved in the Capacitor Core. I currently save all timestamps as ISO 8601 strings.

**Comment by u12206050** [comment]
Any update, this is pretty crucial I would think to any app using firestore?

**Comment by robingenz** [comment]
No, there are no updates yet. It's still on my to-do list for this year. For the time being, I recommend saving all timestamps as ISO 8601 strings.

**Comment by gdrose** [comment]
I'm still facing the same issue. Are there any updates regarding this?

**Comment by beonde** [comment]
Saving the dates as strings to Firestore is not ideal as you will lose querying functionality that is possible with timestamp fields.

In the meantime I have made a local patch package to update the swift files to handle timestamps and output them as ISO 8601 strings from the plugin instead of null. Then it is easy to convert to a javascript date object from there.

The helper struct I have defined in swift looks like this:
```swift
import FirebaseFirestore

struct FirestoreHelper {
    static func convertTimestampsInDocument(data: [String: Any]) -> [String: Any] {
        var convertedData = [String: Any]()
        
        for (key, value) in data {
            if let timestamp = value as? Timestamp {
                // Convert Timestamp to Date
                let date = timestamp.dateValue()
                // Use the Date extension to convert Date to ISO 8601 string
                convertedData[key] = date.toISO8601String()
            } else if let nestedData = value as? [String: Any] {
                // Recursively convert nested dictionaries
                convertedData[key] = convertTimestampsInDocument(data: nestedData)
            } else {
                convertedData[key] = value
            }
        }
        
        return convertedData
    }
}

extension Date {
    func toISO8601String() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0) // Set to UTC for consistency
        return formatter.string(from: self)
    }
}
```
The I was able to modify each toJSObject() method for the Results classes like so:
```swift
public func toJSObject() -> AnyObject {
        // Convert the Firestore timestamps
        let rawData = documentSnapshot.data() ?? [:]
        let convertedData = FirestoreHelper.convertTimestampsInDocument(data: rawData)

        let snapshotDataResult = FirebaseFirestoreHelper.createJSObjectFromHashMap(convertedData)

        var snapshotResult = JSObject()
        snapshotResult["id"] = documentSnapshot.documentID
        snapshotResult["path"] = documentSnapshot.reference.path
        if let snapshotDataResult = snapshotDataResult {
            snapshotResult["data"] = snapshotDataResult
        } else {
            snapshotResult["data"] = NSNull()
        }

        var metadata = JSObject()
        metadata["fromCache"] = documentSnapshot.metadata.isFromCache
        metadata["hasPendingWrites"] = documentSnapshot.metadata.hasPendingWrites
        snapshotResult["metadata"] = metadata

        var result = JSObject()
        result["snapshot"] = snapshotResult
        return result as AnyObject
    }
```
Hope that helps as a temporary solution and maybe can assist towards a permanent solution too.

**Comment by beonde** [comment]
> In poking around the codebase, I've narrowed down the problem to the `FirebaseFirestoreHelper.swift` file. It seems the `createJSValue` method returns `nil` for instances of `Timestamp`. I'm manually marking its type and passing up its value as a second/nano second pair. **_In the JS layer, I'm passing these to the timestamp constructor, making it work just like the web client._** If we could make these changes in `firestore/src/index.ts` and wrap the necessary read/write methods to pass this information back and forth from the native/web layers, that would be great!
> 
> ![Screenshot 2023-10-23 at 4 03 50 PM](https://private-user-images.githubusercontent.com/11188759/277501219-7955a520-a55a-4a8b-8062-f966009c9774.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2MTc4OTgsIm5iZiI6MTczNjYxNzU5OCwicGF0aCI6Ii8xMTE4ODc1OS8yNzc1MDEyMTktNzk1NWE1MjAtYTU1YS00YThiLTgwNjItZjk2NjAwOWM5Nzc0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTExVDE3NDYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyNjlkZTYyMzYzYjM4MTg4NTFhYzFkNDk1Nzc4ZWY2MGM5NWU0M2VjMjY0ODJmMmFkM2ZiMTU2NDZiZGNkMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4K0LWhdwZcX48veM1jm_ipfoYVybXgkfwcYKmldbEuk)

This is actually a much more efficient place to put the conversion. I think I will adjust my implementation to match this. That way you only need to iterate over the fields in a document once instead of twice.

**Comment by trentrand** [comment]
@beonde, can you share a git patch that we can apply in local codebase to the @capacitor-firebase/firestore dependency?

**Comment by cankutaksoy** [comment]
Tested in all 3 platforms, worked like a charm, and after the Capacitor-Firebase officially got fixed it will exit from the second if statement without breaking anything, thank you @trentrand !

I extended your function to a pipe, and since my small app doesn't need the nanosecond part, I just put `.toDate()` inside that pipe:

```typescript
import { Pipe, PipeTransform } from '@angular/core';
import { Timestamp } from 'firebase/firestore';
@Pipe({
  name: 'firebaseTimestampToDate'
})
export class FirebaseTimestampToDatePipe implements PipeTransform {

  transform(value: any): Date | null {
    if (!value) {
      return null;
    }

    if (value instanceof Timestamp) {
      return value.toDate();
    }
    
    // iOS format
    if (value?.type === 'timestamp' && typeof value.seconds === 'string' && typeof value.nanoseconds === 'string') {
      return new Timestamp(
        parseInt(value.seconds, 10),
        parseInt(value.nanoseconds, 10)
      ).toDate();
    }
    
    // Android format
    if (typeof value === 'string' && value.startsWith('Timestamp(')) {
      const match = value.match(/Timestamp\(seconds=(\d+), nanoseconds=(\d+)\)/);
      if (match) {
        return new Timestamp(
          parseInt(match[1], 10),
          parseInt(match[2], 10)
        ).toDate();
      }
    }
    
    if (value instanceof Date) {
      return value;
    }
    
    return null;
  }
}
``` 


So I'm calling it like this:
`{{ event.eventDate | firebaseTimestampToDate | date:'dd MMMM yyyy' }}`

Thanks again.

**Comment by robingenz** [comment]
His PR is great. But I would like to try a different approach first.

**Comment by johnmckay-reward** [comment]
No worries, in the mean time it might be an idea to put a bit of a warning in the docs about features which aren't working

I'm aware that geopoints don't work either, it's a bit frustrating only discovering this once you are well through development :)

**Comment by niteshgrg** [comment]
Thanks for working on this Robin, Any update on this so far. I just started using the plugin and have some Timstamp formated data in db which ideally would like to keep same.

**Comment by kulakowka** [comment]
I apologize if my feedback seemed disrespectful. I wrote this message in the heat of the moment. I was wrong. 😇 😳

I am truly grateful for your work. 

How can I help you solve this problem?

**Comment by robingenz** [comment]
A dev version with support for `Timestamp`, `GeoPoint`, and `FieldValue` is now available for testing:

```
npm i https://pkg.pr.new/capawesome-team/capacitor-firebase/@capacitor-firebase/firestore@972b577
```

We'd love your help testing this before the stable release. Here's how the new types work:

### Reading Timestamps and GeoPoints

`Timestamp` and `GeoPoint` values are now automatically deserialized when reading documents. No extra work needed:

```typescript
import { FirebaseFirestore, Timestamp, GeoPoint } from '@capacitor-firebase/firestore';

const { snapshot } = await FirebaseFirestore.getDocument({ reference: 'posts/abc' });

// snapshot.data.createdAt is now a Timestamp instance
const date = snapshot.data.createdAt.toDate();
const millis = snapshot.data.createdAt.toMillis();

// snapshot.data.location is now a GeoPoint instance
const lat = snapshot.data.location.latitude;
const lng = snapshot.data.location.longitude;
```

### Writing Timestamps and GeoPoints

```typescript
import { FirebaseFirestore, Timestamp, GeoPoint } from '@capacitor-firebase/firestore';

await FirebaseFirestore.setDocument({
  reference: 'posts/abc',
  data: {
    createdAt: Timestamp.now(),
    scheduledFor: Timestamp.fromDate(new Date('2026-06-01')),
    location: new GeoPoint(48.1351, 11.5820),
  },
});
```

### Using FieldValue

```typescript
import { FirebaseFirestore, FieldValue } from '@capacitor-firebase/firestore';

await FirebaseFirestore.updateDocument({
  reference: 'posts/abc',
  data: {
    updatedAt: FieldValue.serverTimestamp(),
    tags: FieldValue.arrayUnion('firebase', 'capacitor'),
    removed: FieldValue.arrayRemove('draft'),
    views: FieldValue.increment(1),
    deprecated: FieldValue.delete(),
  },
});
```

These types work across **all three platforms** (Web, Android, iOS), including in snapshot listeners and query filters.

Please report any issues you encounter in #956. Thanks!
